File: ios/StelliveHubiOS/StelliveHubiOS/Services/NotificationPayload.swift
import Foundation

enum NotificationDeliveryLevel: String, Codable {
    case immediatePush = "immediate_push"
    case summaryPush = "summary_push"
    case inAppHistoryOnly = "in_app_history_only"
}

struct NotificationPayload: Equatable {
    let eventId: String
    let memberId: String
    let generationId: String
    let source: NotificationPlatform
    let eventType: NotificationEventType
    let title: String
    let body: String
    let appDeepLink: String
    let platformUrl: String?
    let deliveryLevel: NotificationDeliveryLevel
    let summaryGroupId: String?
    let supersedesEventIds: [String]

    static func from(data: [String: String]) -> NotificationPayload? {
        guard
            let eventId = nonEmpty(data["eventId"]),
            let memberId = nonEmpty(data["memberId"]),
            let generationId = nonEmpty(data["generationId"]),
            let sourceValue = nonEmpty(data["source"]),
            let source = NotificationPlatform(rawValue: sourceValue),
            let eventTypeValue = nonEmpty(data["eventType"]),
            let eventType = NotificationEventType(rawValue: eventTypeValue),
            let title = nonEmpty(data["title"]),
            let appDeepLink = nonEmpty(data["appDeepLink"])
        else {
            return nil
        }

        return NotificationPayload(
            eventId: eventId,
            memberId: memberId,
            generationId: generationId,
            source: source,
            eventType: eventType,
            title: title,
            body: data["body"] ?? "",
            appDeepLink: appDeepLink,
            platformUrl: nonEmpty(data["platformUrl"]),
            deliveryLevel: NotificationDeliveryLevel(rawValue: data["deliveryLevel"] ?? "") ?? .immediatePush,
            summaryGroupId: nonEmpty(data["summaryGroupId"]),
            supersedesEventIds: data["supersedesEventIds"]?
                .split(separator: ",")
                .map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty } ?? []
        )
    }

    private static func nonEmpty(_ value: String?) -> String? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
            return nil
        }
        return value
    }
}

File: android/StelliveHubAndroid/app/src/main/java/dev/minepacu/stelliveeventnotifier/core/notification/NotificationPayload.kt
package dev.minepacu.stelliveeventnotifier.core.notification

import dev.minepacu.stelliveeventnotifier.core.model.NotificationEventType

enum class NotificationDeliveryLevel(val wireName: String) {
    IMMEDIATE_PUSH("immediate_push"),
    SUMMARY_PUSH("summary_push"),
    IN_APP_HISTORY_ONLY("in_app_history_only");

    companion object {
        fun fromWireName(value: String?): NotificationDeliveryLevel? =
            entries.firstOrNull { it.wireName == value }
    }
}

data class NotificationPayload(
    val eventId: String,
    val memberId: String,
    val generationId: String,
    val source: String,
    val eventType: NotificationEventType,
    val title: String,
    val body: String,
    val appDeepLink: String,
    val platformUrl: String?,
    val deliveryLevel: NotificationDeliveryLevel,
    val summaryGroupId: String?
) {
    companion object {
        fun fromData(data: Map<String, String>): NotificationPayload? {
            val eventType = NotificationEventType.entries.firstOrNull { it.wireName == data["eventType"] } ?: return null
            return NotificationPayload(
                eventId = data["eventId"]?.takeIf { it.isNotBlank() } ?: return null,
                memberId = data["memberId"]?.takeIf { it.isNotBlank() } ?: return null,
                generationId = data["generationId"]?.takeIf { it.isNotBlank() } ?: return null,
                source = data["source"]?.takeIf { it.isNotBlank() } ?: return null,
                eventType = eventType,
                title = data["title"]?.takeIf { it.isNotBlank() } ?: return null,
                body = data["body"] ?: "",
                appDeepLink = data["appDeepLink"]?.takeIf { it.isNotBlank() } ?: return null,
                platformUrl = data["platformUrl"]?.takeIf { it.isNotBlank() },
                deliveryLevel = NotificationDeliveryLevel.fromWireName(data["deliveryLevel"]) ?: NotificationDeliveryLevel.IMMEDIATE_PUSH,
                summaryGroupId = data["summaryGroupId"]?.takeIf { it.isNotBlank() }
            )
        }
    }
}
